mapper/pchain/tx_builder.go

```diff
@@ -218,31 +320,95 @@ func buildAddDelegatorTx(
 		return nil, nil, fmt.Errorf("parse outputs failed: %w", err)
 	}
 
-	memo, err := mapper.DecodeToBytes(metadata.Memo)
+	weight, err := sumOutputAmounts(stakeOutputs)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	tx := &txs.Tx{Unsigned: &txs.AddPermissionlessValidatorTx{
+		BaseTx: txs.BaseTx{BaseTx: avax.BaseTx{
+			NetworkID:    metadata.NetworkID,
+			BlockchainID: blockchainID,
+			Outs:         outs,
+			Ins:          ins,
+		}},
+		Validator: txs.Validator{
+			NodeID: nodeID,
+			Start:  metadata.Start,
+			End:    metadata.End,
+			Wght:   weight,
+		},
+		Subnet:                ids.ID(subnetBytes),
+		Signer:                pop,
+		StakeOuts:             stakeOutputs,
+		ValidatorRewardsOwner: validationRewardsOwner,
+		DelegatorRewardsOwner: delegationRewardsOwner,
+		DelegationShares:      metadata.Shares,
+	}}
+
+	return tx, signers, tx.Sign(codec, nil)
+}
+
+// [buildAddPermissionlessDelegatorTx] returns a duly initialized tx if it does not err
+func buildAddPermissionlessDelegatorTx(
+	matches []*parser.Match,
+	metadata Metadata,
+	codec codec.Manager,
+	avaxAssetID ids.ID,
+) (*txs.Tx, []*types.AccountIdentifier, error) {
+	if metadata.StakingMetadata == nil {
+		return nil, nil, errInvalidMetadata
+	}
+
+	blockchainID := metadata.BlockchainID
+
+	nodeID, err := ids.NodeIDFromString(metadata.NodeID)
 	if err != nil {
-		return nil, nil, fmt.Errorf("parse memo failed: %w", err)
+		return nil, nil, err
+	}
+	rewardsOwner, err := buildOutputOwner(metadata.ValidationRewardsOwners, metadata.Locktime, metadata.Threshold)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ins, _, signers, err := buildInputs(matches[0].Operations, avaxAssetID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("parse inputs failed: %w", err)
+	}
+
+	outs, stakeOutputs, _, err := buildOutputs(matches[1].Operations, codec, avaxAssetID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("parse outputs failed: %w", err)
 	}
 
 	weight, err := sumOutputAmounts(stakeOutputs)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	tx := &txs.Tx{Unsigned: &txs.AddDelegatorTx{
+	subnetBytes, err := mapper.DecodeToBytes(metadata.Subnet)
```

Comment:
Can we default this to `constants.PrimaryNetworkID` if `metadata.Subnet` is undefined / empty? For the time being, primary network is the only one that we are supporting, and I'd prefer not to hardcode empty ids in the client side if possible.